prover/cmd/controller/controller/controller.go

```diff
@@ -31,49 +42,127 @@ func runController(ctx context.Context, cfg *config.Config) {
 		)
 	}
 
-	ctx, stop := signal.NotifyContext(ctx, syscall.SIGTERM)
-	defer stop()
+	// Derive the command context with a cancel function
+	cmdCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	// cmdContext is the context we provide for the command execution. In
-	// spot-instance mode, the context is subordinated to the ctx.
-	cmdContext := context.Background()
-	if cfg.Controller.SpotInstanceMode {
-		cmdContext = ctx
-	}
+	signal.Notify(signalChan, syscall.SIGTERM, syscall.SIGUSR1)
+	defer signal.Stop(signalChan)
+
+	// Signal handler goroutine — race-safe with atomic flags
+	// We never stop listening since it is possible to receive more than one signals
+	// For ex: spot reclaim => SIGUSR1 -> SIGTERM
+	go func() {
+		for sig := range signalChan {
+			switch sig {
+			case syscall.SIGUSR1:
+				spotReclaimDetected.Store(true)
+				if !gracefulShutdownRequested.Load() {
+					cLog.Info("Received SIGUSR1: marking spot reclaim detected, cancelling context ASAP...")
+					cancel()
+					if cfg.Controller.Prometheus.Enabled {
+						metrics.IncSpotInterruption(cfg.Controller.LocalID)
+					}
+				} else {
+					cLog.Info("Received SIGUSR1 after SIGTERM: marking spot reclaim detected (shutdown already in progress)")
+				}
+
+				// Requeue active job immediately for spot reclaim
+				if activeJob != nil {
+					cLog.Infof("Spot reclaim detected. REQUEUE active job: %v", activeJob.OriginalFile)
+					_ = os.Remove(activeJob.TmpResponseFile(cfg))
+					if err := os.Rename(activeJob.InProgressPath(), activeJob.OriginalPath()); err != nil {
+						cLog.Errorf("Failed to requeue job %v: %v", activeJob.InProgressPath(), err)
+					}
+					activeJob = nil
+				}
+			case syscall.SIGTERM:
+				gracefulShutdownRequested.Store(true)
+				if !spotReclaimDetected.Load() {
+					cLog.Info("Received SIGTERM: graceful shutdown requested")
+				}
+				// NOTE: DO NOT call cancel() here to respect the termination grace period
+				// If there's an active job, start a timer to enforce grace period
+				if activeJob != nil {
```

Comment:
if executor.Run() complete right after the check, the Goroutine can sleep 1800s without any job..shall the goroutine re-check if the job is still active after waiting?